R/GraphClasses.R

```diff
@@ -0,0 +1,596 @@
+        }
+
+        , default_node_measures = function(){
+            log_fatal('Default node measures not implemented.')
+        }
+
+        , default_graph_measures = function(){
+            log_fatal('Default graph measures not implemented.')
+        }
+    ) # /active
+
+    , private = list(
+        protected = list(
+            nodes = NULL
+            , edges = NULL
+            , igraph = NULL
+            , graph_measures = list()
+        )
+
+        , initialize_igraph = function(directed){
+
+            log_info("Constructing igraph object...")
+
+            # Connected graph
+            if (nrow(self$edges) > 0) {
+                # A graph with edges
+                connectedGraph <- igraph::graph.edgelist(
+                    as.matrix(self$edges[,list(SOURCE,TARGET)])
+                    , directed = directed
+                )
+            } else {
+                connectedGraph <- igraph::make_empty_graph(directed = directed)
+            }
+
+            # Unconnected graph
+            orphanNodes <- base::setdiff(
+                self$nodes[, node]
+                , unique(c(self$edges[, SOURCE], self$edges[, TARGET]))
+            )
+            unconnectedGraph <- igraph::make_empty_graph(directed = directed) + igraph::vertex(orphanNodes)
+
+            # Complete graph
+            completeGraph <- connectedGraph + unconnectedGraph
+
+            # Store in protected cache
+            private$protected$igraph <- completeGraph
+
+            log_info("...done constructing igraph object.")
+
+            return(invisible(NULL))
+        } # /initialize_igraph
+
+        # Functions for node measures
+        # All functions should return a named vector of node measure values
+        , node_measure_functions = list()
+
+        # Functions for graph-level measures
+        # All functions should return numeric of length 1
+        , graph_measure_functions = list()
+
+    )  # /private
+)
+
+#' @title Directed Graph Network Model
+#' @name DirectedGraph
+#' @description R6 class defining a directed graph model for representing a
+#' network, including methods to calculate various measures from graph theory.
+#' The \link[igraph:igraph-package]{igraph} package is used as a backend for
+#' calculations.
+#' @format An \code{\link[R6]{R6Class}} generator object
+#' @inheritSection GraphClasses Class Constructor
+#' @inheritSection GraphClasses Public Methods
+#' @inheritSection GraphClasses Public Fields
+#' @inheritSection GraphClasses Special Methods
+#' @inheritSection DirectedGraphMeasures Node Measures
+#' @inheritSection DirectedGraphMeasures Graph Measures
+#' @importFrom R6 R6Class
+#' @importFrom igraph degree closeness betweenness
+#' @importFrom igraph page_rank hub_score authority_score
+#' @importFrom igraph neighborhood.size vcount V
+#' @importFrom igraph centralize centr_degree_tmax
+#' @importFrom igraph centr_clo_tmax centr_betw_tmax
```

Comment:
So I for sure want to have a roxygen page for this, whether we export it or not. Because `DirectedGraph` objects are accessible from reporters, there should be documentation about their methods.

I also intentionally did not export this, because I don't anticipate people instantiation them separately from a reporter.